crates/red_knot_python_semantic/src/types/diagnostic.rs

```diff
@@ -233,6 +234,35 @@ declare_lint! {
     }
 }
 
+declare_lint! {
+    /// Checks for `raise` statements that raise non-exceptions or use non-exceptions cause.
+    ///
+    /// ## Why is this bad?
+    /// Only classes inherit from and instances of `BaseException` can be raised.
+    /// Only classes inherit from and instances of `BaseException` and `None`
+    /// can be used as exception cause.
+    /// Attempts to violate this restriction will result in a `TypeError` at runtime.
+    ///
+    /// ## Example
+    /// ```python
+    /// raise int from str()
+    /// ```
+    ///
+    /// Use instead:
+    /// ```python
+    /// raise ZeroDivisionError() from RuntimeError
+    /// ```
```

Comment:
```suggestion
    /// ## Examples
    /// ```python
    /// def f():
    ///     try:
    ///         something()
    ///     except NameError:
    ///         raise "oops!" from f
    ///
    /// def g():
    ///     raise NotImplemented from 42
    /// ```
    ///
    /// Use instead:
    /// ```python
    /// def f():
    ///     try:
    ///         something()
    ///     except NameError as e:
    ///         raise RuntimeError("oops!") from e
    ///
    /// def g():
    ///     raise NotImplementedError from None
    /// ```
```